### PR TITLE
[WIP] Enable Customizable Dist Graph Expansion

### PR DESCRIPTION
**Pull request by mrshenli**

### CODE REVIEW

File: torch/distributed/_spmd/api.py [hunk 1]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+                    return [p.grad for p in mod.parameters()]
+
+            # parameters are at the end of placeholders
+            # FIXME(@mrshenli): let make_fx ignore nn.Module in args? or it should
+            # eliminate unused placeholders?
+            gm = make_fx(partial(stateless_func, func))(
+                args, kwargs, params_and_buffers
+            )
+            gm.graph.print_tabular()
+
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py:234]: I think we need to make the gm caller-flattened here. Otherwise, the expansion passes that remove/allocates addtional parameters won't work.

```python
from typing import Any

import torch
from torch.fx.graph import _PyTreeCodeGen, _PyTreeInfo, CodeGen


class _PyTreeCodeGenOutputsOnly(_PyTreeCodeGen):
    # pyre-ignore[3]
    def process_inputs(self, *args: Any) -> Any:
        return args

    # pyre-ignore[2, 3]
    def gen_fn_def(self, free_vars, maybe_return_annotation):
        return CodeGen.gen_fn_def(self, free_vars, maybe_return_annotation)


def to_caller_flattened_graph_module(gm: torch.fx.GraphModule) -> torch.fx.GraphModule:
    """Move the responsibility of flattening the input arguments from the
    graph module to the caller.

    Example:

        output = gm(my_struct)

        gm = gm(to_caller_flattened_graph_module)

        output = gm(*pytree.flatten(my_struct)[0])
    """
    # pyre-ignore[16]
    gm._graph._codegen = _PyTreeCodeGenOutputsOnly(
        pytree_info=_PyTreeInfo(
            # pyre-ignore[6]
            orig_args=None,
            # pyre-ignore[6]
            in_spec=None,
            # pyre-ignore[16]
            out_spec=gm._graph._codegen.pytree_info.out_spec,
        )
    )
    gm.recompile()
    return gm
```

File: torch/distributed/_spmd/api.py [hunk 2]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+            # FIXME(@mrshenli): support multiple nn.Module instances
+            # FIXME(@mrshenli): do the same for optimizers
+            mod = None
+            for arg in pytree.tree_flatten(list(args) + list(kwargs.values()))[
+                0
+            ]:
+                if isinstance(arg, nn.Module):
+                    mod = arg
+
+            assert mod is not None
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py]: Should we also check that there's only one `nn.Module` argument?

File: torch/distributed/_spmd/api.py [hunk 3]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+            params_and_buffers = {
+                **dict(mod.named_parameters(remove_duplicate=False)),
+                **dict(mod.named_buffers(remove_duplicate=False)),
+            }
+
+            def stateless_func(func, args, kwargs, params_and_buffers):
+                # TODO: do similar thing for optimizer
+                with stateless._reparametrize_module(mod, params_and_buffers):
+                    func(*args, **kwargs)
+                    return [p.grad for p in mod.parameters()]
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py]: If func is a train step, it should consume the gradients right and there's no gradients to return right?

Comment (yifuwang) [inline torch/distributed/_spmd/api.py]: I think we need to be faithful to what func returns. Often times it will return the model forward result which is consumed metric/logging related things and we won't capture these in the compiled training step.

File: torch/distributed/_spmd/api.py [hunk 4]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+
+            # parameters are at the end of placeholders
+            # FIXME(@mrshenli): let make_fx ignore nn.Module in args? or it should
+            # eliminate unused placeholders?
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py:220]: > let make_fx ignore nn.Module in args?

We may be able to do this by currying the nn.Module argument into stateless func.

> or it should eliminate unused placeholders?

IMO it shouldn't do that. A placeholder can be unused right after tracing but used after a pass. For example, if the train step input is `Tuple[torch.Tensor, str]`, after capture, the second placeholder in the graph will always be unused. However, when you call the graph, the `str` value is always passed into the graph, and the passes can leverage that information (`plan_and_shard` leverage this to get the input key orders).

File: torch/distributed/_spmd/api.py [hunk 5]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+
+
+def _compile():
+    def inner(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # 1. Extract nn.Module from args and kwargs
+            # FIXME(@mrshenli): detect nn.Module in a nested way
+            # FIXME(@mrshenli): support multiple nn.Module instances
+            # FIXME(@mrshenli): do the same for optimizers
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py:177]: Can you add optimizer into the example? IMO the purpose of capturing the train step as opposed to just the model fwd/bwd is to capture optimizers. It's not yet clear how things comes together here.

File: torch/distributed/_spmd/api.py [hunk 6]
```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
...
+    def transform(
+        self, gm: fx.GraphModule, context: Dict[fx.Node, DTensor]
+    ) -> fx.Graph:
+        pass
+
+
+OVERRIDES: Dict[Type[Any], Override] = {}
+
+
+def _register_override(typ: Any, override: Override) -> None:
```
Comment (yifuwang) [inline torch/distributed/_spmd/api.py]: I wonder if it's better to pass overrides as arguments to _compile as opposed to registering them. I think most overrides are instances that needs to be customized via configs.

File: torch/distributed/_spmd/api.py (hunk 3 above)
Comment (mrshenli) [reply torch/distributed/_spmd/api.py]: yep, this is temporary before we add optimizer support
